File: Habitica/src/main/java/com/habitrpg/android/habitica/api/HostConfig.java
package com.habitrpg.android.habitica.api;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;

import com.habitrpg.android.habitica.BuildConfig;
import com.habitrpg.android.habitica.R;

/**
 * The configuration of the host<br />
 * Currently, the Port isn't used at all.
 *
 * @author MagicMicky
 */
public class HostConfig {
    private String address;
    private String port;
    private String api;
    private String user;

    public HostConfig(Context context) {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(context);
        this.port = BuildConfig.PORT;
        this.address = BuildConfig.DEBUG ? BuildConfig.BASE_URL : context.getString(R.string.base_url);
        if (prefs.contains("base_url"))
            this.address = prefs.getString("base_url", context.getString(R.string.base_url));
        this.api = prefs.getString("APIToken", null);
        this.user = prefs.getString(context.getString(R.string.SP_userID), null);
    }

    /**
     * Create a new HostConfig
     *
     * @param address the address of the server
     * @param port    the port of the server
     * @param api     the API token of the user
     * @param user    the user ID
     */
    public HostConfig(String address, String port, String api, String user) {
        this.setAddress(address);
        this.setPort(port);
        this.setApi(api);
        this.setUser(user);
    }

    /**
     * @return the address
     */
    public String getAddress() {
        return address;
    }

    /**
     * @param address the address to set
     */
    public void setAddress(String address) {
        this.address = address;
    }

    /**
     * @return the port
     */
    public String getPort() {
        return port;
    }

    /**
     * @param port the port to set
     */
    public void setPort(String port) {
        this.port = port;
    }

    /**
     * @return the api
     */
    public String getApi() {
        return api;
    }

    /**
     * @param api the api to set
     */
    public void setApi(String api) {
        this.api = api;
    }

    /**
     * @return the user
     */
    public String getUser() {
        return user;
    }

    /**
     * @param user the user to set
     */
    public void setUser(String user) {
        this.user = user;
    }

    public boolean hasAuthentication() {
        return user != null && user.length() > 0 && api != null && api.length() > 0;
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/api/HostConfig.kt
package com.habitrpg.android.habitica.api

import android.content.Context
import android.content.SharedPreferences
import android.os.Build
import androidx.core.content.edit
import com.habitrpg.android.habitica.BuildConfig
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.helpers.KeyHelper


/**
 * The configuration of the host<br></br>
 * Currently, the Port isn't used at all.
 *
 * @author MagicMicky
 */
class HostConfig {
    var address: String
    var port: String
    var apiKey: String
    var userID: String

    constructor(sharedPreferences: SharedPreferences, keyHelper: KeyHelper?, context: Context) {
        this.port = BuildConfig.PORT
        if (BuildConfig.DEBUG) {
            this.address = BuildConfig.BASE_URL
        } else {
            val address = sharedPreferences.getString("server_url", null)
            if (address != null && address.isNotEmpty()) {
                this.address = address
            } else {
                this.address = context.getString(R.string.base_url)
            }
        }
        this.userID = sharedPreferences.getString(context.getString(R.string.SP_userID), null) ?: ""

        this.apiKey = loadAPIKey(sharedPreferences, keyHelper)
    }

    private fun loadAPIKey(sharedPreferences: SharedPreferences, keyHelper: KeyHelper?): String {
        return if (sharedPreferences.contains(userID)) {
            val encryptedKey = sharedPreferences.getString(userID, null)
            if (encryptedKey?.isNotBlank() == true) {
                keyHelper?.decrypt(encryptedKey)
            } else {
                ""
            }
        } else {
            val key = sharedPreferences.getString("APIToken", null)
            if (key?.isNotBlank() == true && Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                val encryptedKey = keyHelper?.encrypt(key)
                sharedPreferences.edit {
                    putString(userID, encryptedKey)
                    remove("APIToken")
                }
            }
            key
        } ?: ""
    }

    constructor(address: String, port: String, api: String, user: String) {
        this.address = address
        this.port = port
        this.apiKey = api
        this.userID = user
    }

    fun hasAuthentication(): Boolean {
        return userID.isNotEmpty() && apiKey.isNotEmpty()
    }

}

